server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStore.java

```diff
@@ -210,23 +212,28 @@ else if (map.get(USER_AUTHENTICATION_UAA_PRINCIPAL)!=null) {
         return new OAuth2Authentication(request, userAuthentication);
     }
 
-    protected void performExpirationClean() {
-        Instant last = lastClean.get();
-        //check if we should expire again
-        Instant now = Instant.now();
-        if (enoughTimeHasPassedSinceLastExpirationClean(last, now)) {
-            //avoid concurrent deletes from the same UAA - performance improvement
-            if (lastClean.compareAndSet(last, now)) {
-                try {
-                    JdbcTemplate template = new JdbcTemplate(dataSource);
-                    int expired = template.update(SQL_EXPIRE_STATEMENT, now.toEpochMilli());
-                    logger.debug("[oauth_code] Removed "+expired+" expired entries.");
-                    expired = template.update(SQL_CLEAN_STATEMENT, Timestamp.from(now.minus(LEGACY_CODE_EXPIRATION_TIME)));
-                    logger.debug("[oauth_code] Removed "+expired+" old entries.");
-                } catch (DeadlockLoserDataAccessException e) {
-                    logger.debug("[oauth code] Deadlock trying to expire entries, ignored.");
-                }
+    protected void performExpirationCleanIfEnoughTimeHasElapsed() {
+        if (cleanMutex.tryAcquire()) {
+            //check if we should expire again
+            Instant now = Instant.now();
+            if (enoughTimeHasPassedSinceLastExpirationClean(lastClean, now)) {
+                //avoid concurrent deletes from the same UAA - performance improvement
+                lastClean = now;
+                actuallyPerformExpirationClean(now);
             }
+            cleanMutex.release();
```

Comment:
`cleanMutex.release()` needs to be inside `finally` because it won't be executed when `actuallyPerformExpirationClean()` throws `DataAccessException`, which also should be covered by a unit test case.

Reply:
Yes I was just thinking about putting in a `finally` block.